src/c_kzg_4844.c

```diff
@@ -2773,83 +2773,116 @@ static const fr_t INV_SCALE_FACTOR = {
 };
 
 /**
- * Scale a polynomial in place.
+ * Shift a polynomial in place.
  *
- * Multiplies each coefficient by `1 / scale_factor ^ i`. Equivalent to
- * creating a polynomial that evaluates at `x * k` rather than `x`.
+ * Multiplies each coefficient by `shift_factor ^ i`. Equivalent to
+ * creating a polynomial that evaluates at `x * shift_factor` rather than `x`.
  *
- * @param[in,out]   p       The polynomial coefficients to be scaled
- * @param[in]       len     Length of the polynomial coefficients
+ * @param[in,out]   p            The polynomial coefficients to be scaled
+ * @param[in]       len          Length of the polynomial coefficients
+ * @param[in]       shift_factor Shift factor
  */
-static void scale_poly(fr_t *p, size_t len) {
+static void shift_poly(fr_t *p, size_t len, const fr_t *shift_factor) {
     fr_t factor_power = FR_ONE;
     for (size_t i = 1; i < len; i++) {
-        blst_fr_mul(&factor_power, &factor_power, &INV_SCALE_FACTOR);
+        blst_fr_mul(&factor_power, &factor_power, shift_factor);
         blst_fr_mul(&p[i], &p[i], &factor_power);
     }
 }
 
-/**
- * Unscale a polynomial in place.
+/** Do an FFT over a coset of the roots of unity. The function mutates the
+ *  input array `in`
  *
- * Multiplies each coefficient by `scale_factor ^ i`. Equivalent to creating a
- * polynomial that evaluates at `x / k` rather than `x`.
+ * @param[out]  out The results (array of length n)
+ * @param[in]   in  The input data (array of length n)
+ * @param[in]   n   Length of the arrays
+ * @param[in]   s   The trusted setup
  *
- * @param[in,out]   p       The polynomial coefficients to be unscaled
- * @param[in]       len     Length of the polynomial coefficients
+ * @remark The coset shift factor is SCALE_FACTOR
  */
-static void unscale_poly(fr_t *p, size_t len) {
-    fr_t factor_power = FR_ONE;
-    for (size_t i = 1; i < len; i++) {
-        blst_fr_mul(&factor_power, &factor_power, &SCALE_FACTOR);
-        blst_fr_mul(&p[i], &p[i], &factor_power);
-    }
+static C_KZG_RET coset_fft_fr(
+    fr_t *out, fr_t *in, size_t n, const KZGSettings *s
+) {
+    C_KZG_RET ret;
+
+    shift_poly(in, n, &SCALE_FACTOR);
+
+    /* Compute the FFT */
+    ret = fft_fr(out, in, n, s);
+    if (ret != C_KZG_OK) goto out;
+
+out:
+    return ret;
+}
```

Comment:
Following my suggestion above, so we don't mutate `in`. This is what we would add:
```suggestion
out:
    /* Shift the input back so it's not modified */
    shift_poly(in, n, &INV_SCALE_FACTOR);
    return ret;
}
```

Reply:
I think I prefer the memory allocation approach here: https://github.com/ethereum/c-kzg-4844/pull/449/commits/b815f1d558a123b26b725842320a16390f745d1e#diff-ed568929403e623a4b0c74780ccb41dffc9b594000d41d09b21a407fa2126576R2818-R2839
more than the shift-deshift approach.

This way `in` keeps its `const`ness as well.

Reply:
Reverted to mem alloc in 9751c918